integration/plugin/mountpoint/mountpoint_unix_test.go

```diff
@@ -0,0 +1,866 @@
+// +build !windows
+
+package mountpoint
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/integration-cli/daemon"
+	"github.com/docker/docker/pkg/plugins"
+	"github.com/docker/docker/volume/mountpoint"
+	"github.com/gotestyourself/gotestyourself/skip"
+	"github.com/stretchr/testify/require"
+)
+
+const (
+	testMountPointPlugin = "mountpointplugin"
+	mountFailMessage     = "mount source path contains 'secret'"
+)
+
+var server [6]*httptest.Server
+var d *daemon.Daemon
+var ctrl [6]*mountPointController
+var events []string
+
+type mountPointController struct {
+	propertiesRes mountpoint.PropertiesResponse // propertiesRes holds the plugin response to properties requests
+	attachRes     mountpoint.AttachResponse     // attachRes holds the plugin response to attach requests
+	detachRes     mountpoint.DetachResponse     // detachRes holds the plugin response to detach requests
+	attachCnt     int                           // attachCnt counts the number of attach requests received
+	attachMounts  [][]*mountpoint.MountPoint    // attachMounts is a stack of mount point sets requested for attachment
+}
+
+func setupTest(c *testing.T) func() {
+	testEnv.ProtectImage(c, "busybox:latest")
+
+	d = daemon.New(c, "", dockerdBinary, daemon.Config{
+		Experimental: testEnv.ExperimentalDaemon(),
+	})
+
+	ctrl[0] = &mountPointController{
+		attachRes: mountpoint.AttachResponse{
+			Success: true,
+		},
+		detachRes: mountpoint.DetachResponse{
+			Success: true,
+		},
+		attachCnt:    0,
+		attachMounts: [][]*mountpoint.MountPoint{},
+	}
+	ctrl[1] = &mountPointController{}
+	*ctrl[1] = *ctrl[0]
+	ctrl[2] = &mountPointController{}
+	*ctrl[2] = *ctrl[0]
+	ctrl[3] = &mountPointController{}
+	*ctrl[3] = *ctrl[0]
+	ctrl[4] = &mountPointController{}
+	*ctrl[4] = *ctrl[0]
+	ctrl[5] = &mountPointController{}
+	*ctrl[5] = *ctrl[0]
+
+	typeBind := mountpoint.TypeBind
+	typeVolume := mountpoint.TypeVolume
+
+	// matches -v /host:/container
+	ctrl[0].propertiesRes = mountpoint.PropertiesResponse{
+		Success: true,
+		Patterns: []mountpoint.Pattern{
+			{Type: &typeBind},
+		},
+	}
+	// matches -v /host:/container AND all local volume mounts
```

Comment:
Could you split this out into functions with descriptive names instead of these inline comments?

Then `ctrl` (also needs a more descriptive name) can be created by calling the factory functions.

Reply:
Even better, if all of these aren't used for every test could they be created in the test that uses them?